### PR TITLE
Added node:package sub generator.

### PR DESCRIPTION
**Pull request by hemanth**
@SBoudrias 

`yo node:package` works fine.

But an issue with test cases's `before:all` hooks even though `done()` callback is being called in this test.

P.S: Indentation is screwed up again, duh, will fix them after resolving the test cases.

### CODE REVIEW

File: generators/package/index.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
+'use strict';
+var generators = require('yeoman-generator');
+var npmName = require('npm-name');
+var path = require('path');
+var slugify = require("underscore.string/slugify");
+
+module.exports = generators.Base.extend({
+	initializing: function () {
+		this.fs.copy(
```
Comment (SBoudrias) [inline generators/package/index.js]: Any reason this is occurring first?

File: generators/package/index.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+			this.templatePath('_package.json'),
+			this.destinationPath('package.json')
+		);
+	},
+	askForModuleName: function () {
```
Comment (SBoudrias) [inline generators/package/index.js]: Let's use correct priorities.

File: generators/package/index.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+			name: 'authorUrl',
+			message: 'Author\'s Homepage',
+			store: true
+		}, {
+			name: 'keywords',
+			message: 'Key your keywords (comma to split)'
+		}, {
+			type: 'confirm',
+			name: 'cli',
+			message: 'Do you need a CLI?'
```
Comment (SBoudrias) [inline generators/package/index.js]: Let's delete that option for now.

File: generators/package/index.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+		}, {
+			type: 'confirm',
+			name: 'browser',
+			message: 'Do you need Browserify?'
```
Comment (SBoudrias) [inline generators/package/index.js]: let's delete browserify too.

File: generators/package/index.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+		}];
+
+		this.currentYear = (new Date()).getFullYear();
+
+		this.prompt(prompts, function (props) {
+			if (props.githubUsername) {
+				this.repoUrl = props.githubUsername + '/' + this.slugname;
```
Comment (SBoudrias) [inline generators/package/index.js]: let's store these on `this.props` to keep our user input clean of the generator instance.

File: test/package.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+var helpers = require('yeoman-generator').test;
+
+describe('node:package', function () {
+	before(function (done) {
+		helpers.run(path.join(__dirname, '../generators/package'))
+			.on('end', done);
+	});
+
+	it('creates package.json', function () {
+		assert.file([
```
Comment (SBoudrias) [inline test/package.js]: we need better test here. Let's read and parse the package.json file and make sure the field are correct.

File: generators/package/templates/_package.json [hunk 7]
```diff
@@ -0,0 +1,58 @@
+{
```
Comment (SBoudrias) [inline generators/package/templates/_package.json]: Let's not make this a template. Let's fill values on an object and writeJSON that object.

This way we can extend a pre-existing package.json file and not be destructive.

File: generators/package/index.js (hunk 1 above)
Comment (SBoudrias) [inline generators/package/index.js]: Also, to process a template we need to use `this.fs.copyTpl()` and pass manually the template context we'll be using (as the third parameter).

**Comment by hemanth** [comment]
@SBoudrias Was working on it and noticed:

```
$ DEBUG=yeoman:generator yo node:package
  yeoman:generator Queueing askForModuleName in default +0ms
  yeoman:generator Queueing askFor in default +4ms
  yeoman:generator Queueing initializing in initializing +0ms
  yeoman:generator Running askForModuleName +10ms
? Module Name: meow
? The name above already exists on npm, choose another? No
  yeoman:generator Running initializing +12s
props undefined
  yeoman:generator Running askFor +12ms
? Description: The best module ever.
```

`this.async()` call issue...

File: generators/package/index.js [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+			this.props.keywords = props.keywords.split(',').map(function(el) {
+				return el.trim();
+			});
+
+			done();
+		}.bind(this));
+	},
+  initializing: function () {
+    var pkg = this.fs.readJSON(this.destinationPath('package.json'), {});
+    // We are reaching here even before the askFor is done!
```
Comment (hemanth) [inline generators/package/index.js]: @SBoudrias  at this point we would have only `slugname` and `description` in the context.

File: generators/package/index.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+			message: 'Author\'s Homepage',
+			store: true
+		}, {
+			name: 'keywords',
+			message: 'Key your keywords (comma to split)'
+		}];
+
+		this.currentYear = (new Date()).getFullYear();
+
+		this.prompt(prompts, function (props) {
```
Comment (kevva) [inline generators/package/index.js]: Wrong indent.

File: generators/package/index.js [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+      this.props = props;
```
Comment (kevva) [inline generators/package/index.js]: Wrong indent.

File: generators/package/index.js [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+			}
+
+			this.props.keywords = props.keywords.split(',').map(function(el) {
+				return el.trim();
+			});
+
+			done();
+		}.bind(this));
+	},
+  initializing: function () {
```
Comment (kevva) [inline generators/package/index.js]: Wrong indent.

File: package.json [hunk 12]
```diff
@@ -16,7 +16,8 @@
   },
   "dependencies": {
     "npm-name": "^1.0.2",
-    "yeoman-generator": "^0.19.0"
+    "yeoman-generator": "^0.19.0",
+    "underscore.string": "~3.0.3"
```
Comment (arthurvr) [inline package.json]: Use a caret?

File: generators/package/index.js [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+			}
+		}];
+
+		this.prompt(prompts, function (props) {
+			if (props.pkgName) {
+				return this.askForModuleName();
+			}
+			this.slugname = slugify(props.name);
+			this.safeSlugname = this.slugname.replace(/-+([a-zA-Z0-9])/g, function(
+				g) {
```
Comment (arthurvr) [inline generators/package/index.js]: This seems really weird to me

File: generators/package/index.js [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+		var prompts = [{
+			name: 'name',
+			message: 'Module Name',
+			default: path.basename(process.cwd()),
+		}, {
+			type: 'confirm',
+			name: 'pkgName',
+			message: 'The name above already exists on npm, choose another?',
+			default: true,
+			when: function(answers) {
```
Comment (arthurvr) [inline generators/package/index.js]: Missing space

File: generators/package/index.js [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+				});
+			}
+		}];
+
+		this.prompt(prompts, function (props) {
+			if (props.pkgName) {
+				return this.askForModuleName();
+			}
+			this.slugname = slugify(props.name);
+			this.safeSlugname = this.slugname.replace(/-+([a-zA-Z0-9])/g, function(
```
Comment (arthurvr) [inline generators/package/index.js]: Missing space `function ()`

File: generators/package/index.js [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+		this.prompt(prompts, function (props) {
+      this.props = props;
+			if (props.githubUsername) {
+				this.props.repoUrl = props.githubUsername + '/' + this.slugname;
+			} else {
+				this.props.repoUrl = 'user/repo';
+			}
+
+			this.props.keywords = props.keywords.split(',').map(function(el) {
```
Comment (arthurvr) [inline generators/package/index.js]: Same thing, missing space:  `function () {`

File: generators/package/index.js [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+					}
+					done(false);
+				});
+			}
+		}];
+
+		this.prompt(prompts, function (props) {
+			if (props.pkgName) {
+				return this.askForModuleName();
+			}
```
Comment (kevva) [inline generators/package/index.js]: Add a line break between code blocks.

File: generators/package/index.js [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+var slugify = require("underscore.string/slugify");
+
+module.exports = generators.Base.extend({
+	askForModuleName: function () {
+		var done = this.async();
+
+		var prompts = [{
+			name: 'name',
+			message: 'Module Name',
+			default: path.basename(process.cwd()),
```
Comment (arthurvr) [inline generators/package/index.js]: Trailing comma

**Comment by hemanth** [comment]
@kevva @arthurvr I'm aware of the indentation issues, have mentioned the same in the PR as well, currently focusing on fixing that async issue.

**Comment by kevva** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

File: package.json (hunk 12 above)
Comment (SBoudrias) [inline package.json]: Let's only use Lodash. They have string helpers now.

**Comment by SBoudrias** [comment]
Fix and merged on master.

**Comment by hemanth** [comment]
:+1:

**Pull request closed without merging**